[Merged by Bors] - feat: a complex differentiable function on a disk has a primitive

AlexKontorovich opened the pull request:
We prove that holomorphic functions on discs have primitives. TODO: Extend to holomorphic functions on simply connected domains, and use this to define general complex logarithms.

Co-authored by: Ian Jauslin ian.jauslin@rutgers.edu

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

AlexKontorovich committed:
initial commit

AlexKontorovich committed:
cleaning

AlexKontorovich committed:
HolomorphicOn

AlexKontorovich committed:
lint

AlexKontorovich committed:
moved theorems to appropriate files

AlexKontorovich committed:
added to Mathlib

AlexKontorovich committed:
fixing Set

AlexKontorovich committed:
spacing fix

AlexKontorovich committed:
Update Mathlib/Data/Set/Intervals/UnorderedInterval.lean

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

AlexKontorovich committed:
linter

AlexKontorovich committed:
lint

loefflerd commented on Mathlib/Analysis/Complex/HasPrimitives.lean:
```diff
@@ -0,0 +1,491 @@
+
+/-- *** Moreira's theorem *** A function which is continuous on a disc and whose integral on
+  rectangles in the disc vanishes has a primitive on the disc. -/
+theorem moreiras_theorem {c : ℂ} {r : ℝ} {f : ℂ → ℂ} (f_cont : ContinuousOn f (ball c r))
```
I wonder if this should be given a name reflecting its content, rather than its discoverer? This is the general mathlib convention – e.g., to pick a random example, the Hahn-Banach theorem is called `exists_extension_norm_eq`, not `hahn_banach_theorem`.

AlexKontorovich replied:
```suggestion
theorem VanishesOnRectanglesInDisc.hasPrimitive {c : ℂ} {r : ℝ} {f : ℂ → ℂ} (f_cont : ContinuousOn f (ball c r))
```

AlexKontorovich commented on Mathlib/Analysis/Complex/HasPrimitives.lean:
```diff
@@ -0,0 +1,491 @@
+
+* `deriv_of_wedgeInt`: The derivative of the wedge integral is the function being integrated.
+
+* `moreiras_theorem`: A function which is continuous on a disc and whose integral on
+  rectangles in the disc vanishes has a primitive on the disc (defined by the wedge integral).
```
```suggestion
* `VanishesOnRectanglesInDisc.hasPrimitive`: **Moreira's Theorem**: A function which is 
  continuous on a disc and whose integral on rectangles in the disc vanishes has a primitive
  on the disc (defined by the wedge integral).
```

AlexKontorovich commented on Mathlib/Analysis/Complex/HasPrimitives.lean:
```diff
@@ -0,0 +1,491 @@
+/-- *** Holomorphic functions on discs have Primitives *** A holomorphic function on a disc has
+  primitives. -/
+theorem hasPrimitives_on_disc (c : ℂ) {r : ℝ} : HasPrimitives (ball c r) := fun _ f_holo ↦
+  moreiras_theorem f_holo.continuousOn (vanishesOnRectanglesInDisc_of_holomorphic f_holo)
```
```suggestion
  (vanishesOnRectanglesInDisc_of_holomorphic f_holo).hasPrimitives f_holo.continuousOn
```

AlexKontorovich committed:
Update Mathlib/Analysis/Complex/HasPrimitives.lean

AlexKontorovich committed:
Update Mathlib/Analysis/Complex/HasPrimitives.lean

AlexKontorovich committed:
Update Mathlib/Analysis/Complex/HasPrimitives.lean

AlexKontorovich commented on Mathlib/Analysis/Complex/HasPrimitives.lean:
```diff
@@ -0,0 +1,492 @@
+
+/-- *** Moreira's theorem *** A function which is continuous on a disc and whose integral on
+  rectangles in the disc vanishes has a primitive on the disc. -/
+theorem VanishesOnRectanglesInDisc.hasPrimitive {c : ℂ} {r : ℝ} {f : ℂ → ℂ} (f_cont : ContinuousOn f (ball c r))
+    (hf : VanishesOnRectanglesInDisc c r f) :
```
```suggestion
theorem VanishesOnRectanglesInDisc.hasPrimitive {c : ℂ} {r : ℝ} {f : ℂ → ℂ}
    (f_cont : ContinuousOn f (ball c r)) (hf : VanishesOnRectanglesInDisc c r f) :
```

AlexKontorovich committed:
Update Mathlib/Analysis/Complex/HasPrimitives.lean

AlexKontorovich committed:
Update Mathlib/Analysis/Complex/HasPrimitives.lean

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

AlexKontorovich committed:
name fix

AlexKontorovich committed:
Merge branch 'master' into AK_HasPrimitivesOnDisc

AlexKontorovich commented on Mathlib/Analysis/Complex/CauchyIntegral.lean:
```diff
@@ -584,28 +587,28 @@ Cauchy integral formulas. See also
 differentiable on `s`. -/
-protected theorem _root_.DifferentiableOn.contDiffOn {s : Set ℂ} {f : ℂ → E} {n : ℕ}
-    (hd : DifferentiableOn ℂ f s) (hs : IsOpen s) : ContDiffOn ℂ n f s :=
+protected theorem _root_.DifferentiableOn.contDiffOn {s : Set ℂ} {f : ℂ → E}
```
```suggestion
protected theorem _root_.DifferentiableOn.contDiffOn {s : Set ℂ} {f : ℂ → E} {n : ℕ}
```

AlexKontorovich committed:
Update Mathlib/Analysis/Complex/CauchyIntegral.lean

urkud commented:
I'm going to have a look over the week-end.

AlexKontorovich commented on Mathlib/Analysis/Complex/HasPrimitives.lean:
```diff
@@ -0,0 +1,492 @@
+we prove that holomorphic functions on discs have primitives. The approach is based on Moreira's
+theorem, that a continuous function (on a disc) whose `RectangleIntegral` vanishes on all
+rectangles contained in the disc has a primitive. (Coupled with the fact that holomorphic functions
+satisfy this propoerty.) To prove Moreira's theorem, we first define the `WedgeInt`, which is the
```
```suggestion
satisfy this property.) To prove Morera's theorem, we first define the `WedgeInt`, which is the
```

AlexKontorovich commented on Mathlib/Analysis/Complex/HasPrimitives.lean:
```diff
@@ -0,0 +1,492 @@
+
+* `deriv_of_wedgeInt`: The derivative of the wedge integral is the function being integrated.
+
+* `VanishesOnRectanglesInDisc.hasPrimitive`: **Moreira's Theorem**: A function which is
```
```suggestion
* `VanishesOnRectanglesInDisc.hasPrimitive`: **Morera's Theorem**: A function which is
```

AlexKontorovich commented on Mathlib/Analysis/Complex/HasPrimitives.lean:
```diff
@@ -0,0 +1,492 @@
+
+end WedgeIntDeriv
+
+/-- *** Moreira's theorem *** A function which is continuous on a disc and whose integral on
```
```suggestion
/-- *** Morera's theorem *** A function which is continuous on a disc and whose integral on
```

AlexKontorovich committed:
Update Mathlib/Analysis/Complex/HasPrimitives.lean

AlexKontorovich committed:
Update Mathlib/Analysis/Complex/HasPrimitives.lean

AlexKontorovich committed:
Update Mathlib/Analysis/Complex/HasPrimitives.lean

AlexKontorovich committed:
Update Mathlib/Analysis/Complex/HasPrimitives.lean

urkud commented:
I'm very sorry for the delay. I'll review tonight or tomorrow in the morning.

AlexKontorovich commented:
No rush! I know you're doing a million things at once...

urkud commented on Mathlib/Analysis/Complex/HasPrimitives.lean:
```diff
@@ -0,0 +1,492 @@
+section Rectangle
+
+/-- A `Rectangle` is an axis-parallel rectangle with corners `z` and `w`. -/
+def Rectangle (z w : ℂ) : Set ℂ := [[z.re, w.re]] ×ℂ [[z.im, w.im]]
```
If we introduce this definition, then it should be near `reProdIm` (if `uIcc` is available there), and existing code should be migrated to it.

AlexKontorovich replied:
ok, attempting refactor...

urkud commented:
It should be possible to prove the main theorem for any function `f : Complex -> E`, where `E` is a complex Banach space.

AlexKontorovich committed:
attempted refactoring

AlexKontorovich committed:
fixes

AlexKontorovich committed:
fixes

AlexKontorovich commented:
> It should be possible to prove the main theorem for any function `f : Complex -> E`, where `E` is a complex Banach space.

Ok, should be done now. Thanks!

AlexKontorovich committed:
lint

sgouezel commented on Mathlib/Analysis/Complex/HasPrimitives.lean:
```diff
@@ -0,0 +1,435 @@
+
+open Complex Topology Set Metric
+
+set_option autoImplicit true
```
Could you remove this? Mostly, it confuses people.

AlexKontorovich replied:
Hmm why did I think I needed that? Looks like I don't. It's gone. Thanks!

AlexKontorovich committed:
kill autoimplicit

urkud commented:
I think about proving Poincaré theorem for any smoothly contractible domain, then this will be an easy corollary. If I don't do this by this Sunday, then I'll review the details of your proof.

AlexKontorovich commented:
> I think about proving Poincaré theorem for any smoothly contractible domain, then this will be an easy corollary. If I don't do this by this Sunday, then I'll review the details of your proof.

Any luck? Thanks!

AlexKontorovich committed:
Merge branch 'master' into AK_HasPrimitivesOnDisc

AlexKontorovich committed:
Merge branch 'master' into AK_HasPrimitivesOnDisc

AlexKontorovich committed:
elaboration order problem fixed

AlexKontorovich committed:
fix

AlexKontorovich committed:
fixes

grunweg commented:
Coming here from PR triage: it seems this is waiting for review by @urkud.
@AlexKontorovich In the mean time, can you merge master, please? Reviewing code which applies to current mathlib is always nicer... Thanks!

vbeffara commented:
To show existence of primitives on balls, is seems that using `Complex.hasSum_taylorSeries_on_ball` and constructing a formal primitive of the Taylor series would be quicker than going through Morera (the way I learned it went through Morera -> holomorphic functions are C1 -> Cauchy formula -> everything else, the first step of which is matches this PR, but that feels a little like reinventing things that are in Mathlib already).

OTOH I know that rectangle integration is fundamental in PNT+ so it makes sense to have API for it in Mathlib.

AlexKontorovich commented:
> To show existence of primitives on balls, is seems that using `Complex.hasSum_taylorSeries_on_ball` and constructing a formal primitive of the Taylor series would be quicker than going through Morera (the way I learned it went through Morera -> holomorphic functions are C1 -> Cauchy formula -> everything else, the first step of which is matches this PR, but that feels a little like reinventing things that are in Mathlib already).

Yeah, that's a good point. I'm happy to abandon this approach entirely and switch to Taylor series. (This was a good warmup for understanding how to do PNT, but may no longer be necessary, having served its purpose...)

> OTOH I know that rectangle integration is fundamental in PNT+ so it makes sense to have API for it in Mathlib.

I wouldn't say that it's "fundamental" (in theory, we should eventually have the Jordan curve theorem, and be able to work more like the natural language proof...), but certainly it's proved useful to move things along...

vbeffara commented:
> I wouldn't say that it's "fundamental" (in theory, we should eventually have the Jordan curve theorem, and be able to work more like the natural language proof...), but certainly it's proved useful to move things along...

Sure, let's say essential to the current proof then :-)

AlexKontorovich committed:
Merge branch 'master' into AK_HasPrimitivesOnDisc

AlexKontorovich committed:
Update Mathlib.lean

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

ocfnash committed:
merge master

ocfnash committed:
Fix after merging master

ocfnash committed:
Style nits

ocfnash committed:
Fix `Mathlib.lean`

ocfnash committed:
Some golf

AlexKontorovich commented on Mathlib/Analysis/Complex/HasPrimitives.lean:
```diff
@@ -0,0 +1,448 @@
+
+In this file, we give conditions under which holomorphic functions have primitives. The main goal
+is to prove that holomorphic functions on simply connected domains have primitives. As a first step,
+we prove that holomorphic functions on discs have primitives. The approach is based on Morera's
```
sure, happy to go with disk

AlexKontorovich commented:
Oh I see... I'm no longer able to make changes to this, because of the new PR protocol (but Oliver was able to, of course, since he's a maintainer). Ugh. Can you please remind me the process for moving PRs to forks, without losing all the comments? Anyway, I agree to all the suggested changes; I just don't have the option to accept them myself... Thanks!

ocfnash committed:
Sebastien's suggestions

ocfnash committed:
Sebastien's suggestions (part II, the sequel)

ocfnash commented:
> Oh I see... I'm no longer able to make changes to this, because of the new PR protocol (but Oliver was able to, of course, since he's a maintainer). Ugh. Can you please remind me the process for moving PRs to forks, without losing all the comments? Anyway, I agree to all the suggested changes; I just don't have the option to accept them myself... Thanks!

@AlexKontorovich The new workflow is to create from a fork, clone that locally, do all your work in this new local clone, and then PR from a branch of that repo into the master branch on the real Mathlib. We actually have [a script](https://github.com/leanprover-community/mathlib4/blob/master/scripts/migrate_to_fork.py) that will create the fork AND migrate an existing PR like this into a new PR from the fork.

Even maintainers who have retained write access to Mathlib should and do follow this workflow.

However I was thinking of adopting this PR for you and making an exception in this case. Would that help, or would you prefer to continue working with this PR yourself?

ocfnash committed:
Rename remaining `disc` to `disk`

AlexKontorovich commented:
If you were willing to adopt it, that would be amazing, thanks! Please let me know if there's anything I should do from my end.

ocfnash commented:
> If you were willing to adopt it, that would be amazing, thanks! Please let me know if there's anything I should do from my end.

OK great, I'll get this over the line!

ocfnash committed:
Bit more:
 * Naming tweaks
 * Add `TODO` about `HasPrimitives` (I'll resolve this in due course)
 * Add lemma connecting "wedge" and "rectange" integrals (should be helpful below)

ocfnash committed:
Final round of `Disc` to `Disk` (previously missed these because of capitalisation)

Also a minor style tweak to `VanishesOnRectanglesInDisk`

ocfnash committed:
Golfing

ocfnash committed:
Bit more

ocfnash committed:
Refactoring

I haven't yet worked on the first ~200 lines at all. Even where
I have worked, the naming still needs work (I'm not even sure about
some of the names I have proposed).

I have to stop for now though.

ocfnash committed:
Move lemma + generalise

ocfnash committed:
Improve naming

Still not sure but this is less bad than what I had

ocfnash committed:
Drop `rectangleIntegral`: we don't need this definition.

I really must stop now!

ocfnash committed:
Bit more

ocfnash committed:
And more

ocfnash committed:
Merge branch 'master' into AK_HasPrimitivesOnDisc

ocfnash committed:
Getting there

ocfnash committed:
Relax `NormedRing E` to `NormedAddCommGroup E`

ocfnash committed:
Some more golf

ocfnash committed:
Inline extremely-specific continuity lemmas

These were all extremely specific and used only once.

ocfnash committed:
Begin tidy up of remaining auxiliary lemmata

I may have to step back and ask why we need so many adhoc lemmas

In addition, it may be that for those auxiliary lemmas used only once,
they may no longer be necessary if we improve the proof which
currently demands them.

ocfnash committed:
Golf a bit

OK better stop now

ocfnash committed:
Minor tweaks

ocfnash commented:
I think this PR is ready for review now. Points I see for discussion are:

1. Do we definitely want to introduce the `abbrev HolomorphicOn`? If so we'll also have to have `Holomorphic`, `HolomorphicAt` and potentially have issues with dot notation.
2. Are the names `Complex.IsClosedOn` and `Complex.IsExactOn` (suggested by me) desirable?
3. Are we interested in golfing the `private` auxiliary lemmas?

ocfnash committed:
Gather `variable`s

ocfnash committed:
Back out `abbrev HolomorphicOn`

We may still wish to introduce something along these lines (e.g., some
notation) but it should happen as a separate PR.

mattrobball reviewed (commented):
Other than this, looks good!

ocfnash committed:
CR suggestion: rename `IsClosedOn` to `IsConservativeOn`

ocfnash committed:
Use lowercase where appropriate

jcommelin reviewed (commented):
Thanks :tada:

bors merge
